model_interfaces/charlie_interface.py

```diff
@@ -1,143 +1,149 @@
+            self.display_name = settings.get("display_name", [self.username])[0]
+            self.initial_costs_usd = settings.get("usage", {}).get("total_cost", 0.0)
+        else:
+            raise ValueError("Failed to load initial settings.")
 
-        body = {
+    def reply(self, user_message: str, agent_response: Optional[str] = None) -> str:
+        message_url = f"{self.endpoint}/message/"
+        message_data = {
             "prompt": user_message,
-            "username": self.user_name,
             "display_name": self.display_name,
             "chat_id": self.chat_id,
+            "username": self.username,
         }
-
-        response_json = json.loads(
-            requests.post(self.endpoint + "/message/", headers=headers, json=body).text
-        )
-
-        # Update costs
-        settings = self.get_settings()
-        self.costs_usd = settings["usage"]["total_cost"] - self.initial_costs_usd
-
-        try:
+        response = self.session.post(message_url, json=message_data)
+        if response.status_code == 200:
+            response_json = response.json()
+            self.update_costs()
             return response_json["content"]
-        except KeyError as exc:
-            exc.add_note(f"Received JSON:\n{response_json}")
-            raise
-
-    def get_settings(self):
-        headers = {
-            "Content-Type": "application/json",
-            "Cookie": f"session_token={self.token}",
-        }
-        body = {"username": self.user_name}
-
-        settings = requests.post(
-            self.endpoint + "/load_settings/", headers=headers, json=body
-        )
-        return json.loads(settings.text)
+        else:
+            raise ValueError("Failed to send message.")
+
+    def update_costs(self):
+        load_settings_url = f"{self.endpoint}/load_settings/"
+        response = self.session.post(load_settings_url)
+        if response.status_code == 200:
+            settings = response.json()
+            current_cost = settings.get("usage", {}).get("total_cost", 0.0)
+            self.costs_usd = current_cost - self.initial_costs_usd
+        else:
+            raise ValueError("Failed to update costs.")
 
     def reset(self):
-        # Delete the user data
-        headers = {
-            "Content-Type": "application/json",
-            "Cookie": f"session_token={self.token}",
-        }
-        body = {"username": self.user_name}
-        delete_req = requests.post(
-            self.endpoint + "/delete_data_keep_settings", headers=headers, json=body
-        )
-
+        delete_data_url = f"{self.endpoint}/delete_data_keep_settings/"
+        response = self.session.post(delete_data_url)
+        if response.status_code != 200:
+            raise ValueError("Failed to reset user data.")
 
     def load(self):
         # Charlie mnemonic is web based and so doesn't need to be manually told to resume a conversation
+        # Use the chat_id to continue a conversation
```

Comment:
I'd avoid resuming with charlie until we have a checkpoint-based persistence mechanism. The benchmark relies on setting the agent back to a specific point, while the web's state might be anywhere between two checkpoints.